src/bin/report_collector.rs

```diff
@@ -0,0 +1,383 @@
+    field: FieldType,
+}
+
+impl From<&CommandInput> for InputSource {
+    fn from(source: &CommandInput) -> Self {
+        if let Some(ref file_name) = source.input_file {
+            InputSource::from_file(file_name)
+        } else {
+            InputSource::from_stdin()
+        }
+    }
+}
+
+#[derive(Debug, Subcommand)]
+enum TestAction {
+    /// Execute IPA in semi-honest honest majority setting
+    SemiHonestIpa(IpaQueryConfig),
+    /// Execute IPA in malicious honest majority setting
+    MaliciousIpa(IpaQueryConfig),
+    /// Generate inputs for IPA
+    GenIpaInputs {
+        /// Number of records to generate
+        #[clap(long, short = 'n')]
+        count: u32,
+
+        /// The destination file for generated records
+        #[arg(long)]
+        output_file: Option<PathBuf>,
+
+        #[clap(flatten)]
+        gen_args: EventGeneratorConfig,
+    },
+}
+
+#[derive(Debug, clap::Args)]
+struct GenInputArgs {
+    /// Maximum records per user
+    #[clap(long)]
+    max_per_user: u32,
+    /// number of breakdowns
+    breakdowns: u32,
+}
+
+async fn clients_ready(clients: &[MpcHelperClient; 3]) -> bool {
+    clients[0].echo("").await.is_ok()
+        && clients[1].echo("").await.is_ok()
+        && clients[2].echo("").await.is_ok()
+}
+
+fn validate<I, S>(expected: I, actual: I)
+where
+    I: IntoIterator<Item = S>,
+    I::IntoIter: ExactSizeIterator,
+    S: PartialEq + Debug,
+{
+    let mut expected = expected.into_iter().fuse();
+    let mut actual = actual.into_iter().fuse();
+    let mut mismatch = Vec::new();
+
+    let mut table = Table::new();
+    table.set_header(vec!["Row", "Expected", "Actual", "Diff?"]);
+
+    let mut i = 0;
+    loop {
+        let next_expected = expected.next();
+        let next_actual = actual.next();
+
+        if next_expected.is_none() && next_actual.is_none() {
+            break;
+        }
+
+        let same = next_expected == next_actual;
+        let color = if same { Color::Green } else { Color::Red };
+        table.add_row(vec![
+            Cell::new(format!("{}", i)).fg(color),
+            Cell::new(format!("{:?}", next_expected)).fg(color),
+            Cell::new(format!("{:?}", next_actual)).fg(color),
+            Cell::new(if same { "" } else { "X" }),
+        ]);
+
+        if !same {
+            mismatch.push((i, next_expected, next_actual))
+        }
+
+        i += 1;
+    }
+
+    tracing::info!("\n{table}\n");
+
+    assert!(
+        mismatch.is_empty(),
+        "Expected and actual results don't match: {:?}",
+        mismatch
+    );
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    let args = Args::parse();
+    let _handle = args.logging.setup_logging();
+
+    let scheme = if args.disable_https {
+        Scheme::HTTP
+    } else {
+        Scheme::HTTPS
+    };
+
+    let network_path = args.network.as_deref();
```

Comment:
this can be combined with `test_mpc` logic to initialize clients - maybe just a helper function inside the `cli` module?

Reply:
good point, let me do that